keps/draft-20180426-namespace-template.md

```diff
@@ -0,0 +1,513 @@
+  apiGroup: rbac.authorization.k8s.io
+subjects:
+- kind: User
+  name: "user1"
+  apiGroup: rbac.authorization.k8s.io
+```
+
+This approach changes the namespace defaulting across the cluster: by default (without RBAC), nothing will be populated.
+Both approaches can exist at the same time, with annotation on the `NamespaceTemplate` to decide which approach to go.
+
+```
+policy/namespace-template-activation=rbac
+```
+
+To select the RBAC based approach.
+
+MVP scope only covers `labelSelector` based approach.
+
+#### Apply the template
+
+The list of `templates` in `NamespaceTemplate` is extracted and concatenated as multi-document YAML
+(separated by `---` line between two objects).
+The labels/annotations are injected from `common` during this process with additionally
+
+```
+policy/namespace-template-name=<name of NamespaceTemplate>
+```
+
+To indicate which namespace template creates this object.
+
+Then `kubectl apply` is used to apply the template.
+The reason to use `kubectl` is that the apply logic (3-way merge) is very complicated,
+and currently performed by `kubectl` (not in API server).
+Once the logic is available in API server, the APIs will be used directly.
+`--prune` options is used together with `-l` to clean up old objects no longer defined when namespace template changes.
+
+When multiple templates matches the same namespace, 
+the templates are concatenated in the alphabetic order of the name of `NamespaceTemplate`.
+
+A `NamespaceTemplate` can be disabled by an explicit annotation:
+
+```
+policy/namespace-template-apply=disable
+```
+
+#### Schema validation
+
+As `NamespaceTemplate` is defined as CRD (Custom Resource Definition),
+currently there's no effective way to validate the schema of CR (Custom Resource)
+if it contains complex types, or encapsulating existing object types.
+This problem is out-of-scope here.
+As a result, the `templates` are not validated during the creation of `NamespaceTemplate`,
+and it will fail at population time (via `kubectl apply`) if anything is written incorrectly.
+The failure will be reported as Kubernetes Events in the target namespace.
+
+An alternative is to create a template namespace,
+and put all these objects into the namespace as real resources.
+When a new namespace is created, the objects in the template namespace are copied over.
+As real objects are created, they are effective, and may cause side effects,
+though within known object types, there's no impact of the effectiveness of these objects.
+
+There are other alternatives that the objects can be defined somewhere else from `NamespaceTemplate`.
+The implementation can be pluggable (e.g. adding a `source` property) to support different sources of templates.
+While the initial proposal will focus on inlined objects.
+
+The MVP scope covers the inlining templates only.
+
+#### Opt-out
+
+Cluster administrator is able to opt-out some namespaces from being populated by the controller.
```

Comment:
I don't think the mechanism described here is going to fit well in existing clusters. 

Since this mechanism does nothing to rate limit or control namespace creation under quota, your proposed solution doesn't really solve "defend against malicious creators", so I don't think namespace template should default to covering all namespaces or attempt to solve that problem.  Instead, this proposal is about "automated defaulting", and so it needs to be opt in by the user specifying a label.  If the administrator sets a namespace template label that intersects with a system namespace, that's the administrator's fault.

Reply:
Agree with opt-in

Reply:
I may be misunderstanding, but I believe that making this opt-in runs counter to the goal here, which is to allow the cluster admin to force policies on new namespaces.

Of course obviously it's opt-in in the sense that if the cluster admin decides not to use the feature, then nothing happens, i.e. no change from today. 

It seems that your main objection to this proposal is that it does not address quota for namespace creation. Am I understanding correctly?